flashinfer/fused_moe/utils.py

```
@@ -192,27 +192,102 @@ def nearest_in_buckets(x: int, buckets: List[int]) -> int:
-    return tuple(num_token_buckets)
+    """Generate tuning buckets with adaptive spacing.
+
+    Pure power-of-2 spacing creates huge gaps at large values (e.g. 1024
+    between bucket 1024 and 2048).  For MoE workloads the
+    avg_tokens_per_expert can jump across multiple tile boundaries inside a
+    single gap, forcing the autotuner to pick a kernel optimised for a very
+    different workload size.
+
+    This function uses four phases with progressively coarser spacing::
+
+        Phase 1:  [min .. 256]   — power-of-2    (step ×2)
```

Ruff (0.15.10)

[warning] Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)